{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python -c \"import pykeen\" || pip install git+https://github.com/pykeen/pykeen.git\n",
    "pykeen.env()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/ferran/PhD/yaleCollab\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n",
    "base_dir = f\"{os.getcwd()}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore PyKEEN and default data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.9/site-packages/pykeen/datasets/nations/train.txt\n",
      "TriplesFactory(num_entities=14, num_relations=55, create_inverse_triples=False, num_triples=1592, path=\"/usr/local/lib/python3.9/site-packages/pykeen/datasets/nations/train.txt\")\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pykeen.triples import TriplesFactory\n",
    "from pykeen.pipeline import pipeline\n",
    "from pykeen.datasets.nations import NATIONS_TRAIN_PATH\n",
    "\n",
    "tf = TriplesFactory.from_path(NATIONS_TRAIN_PATH)\n",
    "\n",
    "print(NATIONS_TRAIN_PATH)\n",
    "print(tf)\n",
    "\n",
    "# training, testing = tf.split()\n",
    "\n",
    "# result = pipeline(\n",
    "\n",
    "#     training=training,\n",
    "\n",
    "#     testing=testing,\n",
    "\n",
    "#     model='TransE',\n",
    "\n",
    "#     epochs=5,  # short epochs for testing - you should go higher\n",
    "\n",
    "# )\n",
    "\n",
    "# result.save_to_directory('doctests/test_unstratified_transe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>netherlands</th>\n",
       "      <th>militaryalliance</th>\n",
       "      <th>uk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>egypt</td>\n",
       "      <td>intergovorgs3</td>\n",
       "      <td>usa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>jordan</td>\n",
       "      <td>relbooktranslations</td>\n",
       "      <td>usa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>poland</td>\n",
       "      <td>timesincewar</td>\n",
       "      <td>ussr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>uk</td>\n",
       "      <td>negativebehavior</td>\n",
       "      <td>ussr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>poland</td>\n",
       "      <td>relintergovorgs</td>\n",
       "      <td>uk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586</th>\n",
       "      <td>cuba</td>\n",
       "      <td>ngoorgs3</td>\n",
       "      <td>usa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>usa</td>\n",
       "      <td>timesincewar</td>\n",
       "      <td>china</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>china</td>\n",
       "      <td>treaties</td>\n",
       "      <td>burma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1589</th>\n",
       "      <td>netherlands</td>\n",
       "      <td>eemigrants</td>\n",
       "      <td>usa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>indonesia</td>\n",
       "      <td>embassy</td>\n",
       "      <td>egypt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1591 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      netherlands     militaryalliance     uk\n",
       "0           egypt        intergovorgs3    usa\n",
       "1          jordan  relbooktranslations    usa\n",
       "2          poland         timesincewar   ussr\n",
       "3              uk     negativebehavior   ussr\n",
       "4          poland      relintergovorgs     uk\n",
       "...           ...                  ...    ...\n",
       "1586         cuba             ngoorgs3    usa\n",
       "1587          usa         timesincewar  china\n",
       "1588        china             treaties  burma\n",
       "1589  netherlands           eemigrants    usa\n",
       "1590    indonesia              embassy  egypt\n",
       "\n",
       "[1591 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('/usr/local/lib/python3.9/site-packages/pykeen/datasets/nations/train.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Format data for PYKEEN is in the form of a TSV file with the (h,r,t) format.\n",
    "THis means that the edges are indeed directed.\n",
    "\n",
    "WEe can now take either one of two approaches:\n",
    "\n",
    "- Use STRING database ass a proof of concept. All of these interactions will be considered as @signsallnig interactions. KG will end up having 2 types of edges: those from strign and those from cellchat\n",
    "- Proper approach with the Functional INteraction network from reactome. Here interaction edges can be of different types, and string comprehension will have to be built to decode the direction column into individual triplets with the (h,r,t) format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RPY2 cheatsheet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using R from Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1mrpy2 version:\u001b[0m\n",
      "3.4.5\n",
      "\u001b[1mPython version:\u001b[0m\n",
      "3.9.13 (main, Aug  7 2022, 01:33:23) \n",
      "[Clang 13.1.6 (clang-1316.0.21.2.5)]\n",
      "\u001b[1mLooking for R's HOME:\u001b[0m\n",
      "    Environment variable R_HOME: /usr/local/Cellar/r/4.2.1_2/lib/R\n",
      "    Calling `R RHOME`: /usr/local/Cellar/r/4.2.1_2/lib/R\n",
      "    Environment variable R_LIBS_USER: None\n",
      "\u001b[1mR's additions to LD_LIBRARY_PATH:\u001b[0m\n",
      "/usr/local/lib/R/library/stats/libs/\n",
      "\u001b[1mR version:\u001b[0m\n",
      "    In the PATH: R version 4.2.1 (2022-06-23) -- \"Funny-Looking Kid\"\n",
      "    Loading R library from rpy2: OK\n",
      "\u001b[1mAdditional directories to load R packages from:\u001b[0m\n",
      "None\n",
      "\u001b[1mC extension compilation:\u001b[0m\n",
      "  include:\n",
      "  ['/usr/local/Cellar/r/4.2.1_2/lib/R/include']\n",
      "  libraries:\n",
      "  ['R', 'pcre2-8', 'lzma', 'bz2', 'z', 'icucore', 'dl', 'm', 'iconv']\n",
      "  library_dirs:\n",
      "  ['/usr/local/opt/gettext/lib', '/usr/local/opt/readline/lib', '/usr/local/opt/xz/lib', '/usr/local/lib', '/usr/local/opt/gettext/lib', '/usr/local/opt/readline/lib', '/usr/local/opt/xz/lib', '/usr/local/lib', '/usr/local/Cellar/r/4.2.1_2/lib/R/lib', '/usr/local/Cellar/pcre2/10.40/lib']\n",
      "  extra_compile_args:\n",
      "  []\n",
      "  extra_link_args:\n",
      "  []\n"
     ]
    }
   ],
   "source": [
    "import rpy2\n",
    "import rpy2.situation\n",
    "for row in rpy2.situation.iter_info(): # Print Rpy2 info\n",
    "    print(row)\n",
    "\n",
    "from rpy2.robjects.packages import importr\n",
    "base = importr(\"base\")\n",
    "here = importr(\"here\")\n",
    "\n",
    "if here.here()[0] != base_dir:\n",
    "    raise ValueError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2.robjects.packages import importr\n",
    "base = importr(\"base\")\n",
    "here = importr(\"here\")\n",
    "\n",
    "INTepi21 = base.readRDS(here.here(\"CRCTME_story/Integration_CRCTME-Aug21/RData\",\n",
    "\"INTepi21_filt_2022.rds\"))\n",
    "INTepi21"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load R chunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "R.version.string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i INTepi21\n",
    "library(base)\n",
    "library(tidyverse)\n",
    "library(Seurat)\n",
    "library(here)\n",
    "\n",
    "INTepi21"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STRING + CellchatDB KGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1mrpy2 version:\u001b[0m\n",
      "3.4.5\n",
      "\u001b[1mPython version:\u001b[0m\n",
      "3.9.13 (main, Aug  7 2022, 01:33:23) \n",
      "[Clang 13.1.6 (clang-1316.0.21.2.5)]\n",
      "\u001b[1mLooking for R's HOME:\u001b[0m\n",
      "    Environment variable R_HOME: /usr/local/Cellar/r/4.2.1_2/lib/R\n",
      "    Calling `R RHOME`: /usr/local/Cellar/r/4.2.1_2/lib/R\n",
      "    Environment variable R_LIBS_USER: None\n",
      "\u001b[1mR's additions to LD_LIBRARY_PATH:\u001b[0m\n",
      "/usr/local/lib/R/library/stats/libs/\n",
      "\u001b[1mR version:\u001b[0m\n",
      "    In the PATH: R version 4.2.1 (2022-06-23) -- \"Funny-Looking Kid\"\n",
      "    Loading R library from rpy2: OK\n",
      "\u001b[1mAdditional directories to load R packages from:\u001b[0m\n",
      "None\n",
      "\u001b[1mC extension compilation:\u001b[0m\n",
      "  include:\n",
      "  ['/usr/local/Cellar/r/4.2.1_2/lib/R/include']\n",
      "  libraries:\n",
      "  ['R', 'pcre2-8', 'lzma', 'bz2', 'z', 'icucore', 'dl', 'm', 'iconv']\n",
      "  library_dirs:\n",
      "  ['/usr/local/opt/gettext/lib', '/usr/local/opt/readline/lib', '/usr/local/opt/xz/lib', '/usr/local/lib', '/usr/local/opt/gettext/lib', '/usr/local/opt/readline/lib', '/usr/local/opt/xz/lib', '/usr/local/lib', '/usr/local/Cellar/r/4.2.1_2/lib/R/lib', '/usr/local/Cellar/pcre2/10.40/lib']\n",
      "  extra_compile_args:\n",
      "  []\n",
      "  extra_link_args:\n",
      "  []\n"
     ]
    }
   ],
   "source": [
    "import rpy2\n",
    "import rpy2.situation\n",
    "for row in rpy2.situation.iter_info(): # Print Rpy2 info\n",
    "    print(row)\n",
    "\n",
    "from rpy2.robjects.packages import importr\n",
    "base = importr(\"base\")\n",
    "here = importr(\"here\")\n",
    "\n",
    "if here.here()[0] != base_dir:\n",
    "    raise ValueError\n",
    "\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "pandas2ri.activate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download cellchat DB with the following:\n",
    "\n",
    "`wget https://github.com/sqjin/CellChat/raw/master/data/CellChatDB.mouse.rda`\n",
    "\n",
    "TO use in the cloud, ahve logic that looks for local file, if not present, try to download with ipynb cloud !bash notation and if it fails (for reasons or because user is local) just prompt to download from URL."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_name</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "      <th>agonist</th>\n",
       "      <th>antagonist</th>\n",
       "      <th>co_A_receptor</th>\n",
       "      <th>co_I_receptor</th>\n",
       "      <th>evidence</th>\n",
       "      <th>annotation</th>\n",
       "      <th>interaction_name_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TGFB1_TGFBR1_TGFBR2</th>\n",
       "      <td>TGFB1_TGFBR1_TGFBR2</td>\n",
       "      <td>TGFb</td>\n",
       "      <td>Tgfb1</td>\n",
       "      <td>TGFbR1_R2</td>\n",
       "      <td>TGFb agonist</td>\n",
       "      <td>TGFb antagonist</td>\n",
       "      <td></td>\n",
       "      <td>TGFb inhibition receptor</td>\n",
       "      <td>KEGG: mmu04350</td>\n",
       "      <td>Secreted Signaling</td>\n",
       "      <td>Tgfb1 - (Tgfbr1+Tgfbr2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB2_TGFBR1_TGFBR2</th>\n",
       "      <td>TGFB2_TGFBR1_TGFBR2</td>\n",
       "      <td>TGFb</td>\n",
       "      <td>Tgfb2</td>\n",
       "      <td>TGFbR1_R2</td>\n",
       "      <td>TGFb agonist</td>\n",
       "      <td>TGFb antagonist</td>\n",
       "      <td></td>\n",
       "      <td>TGFb inhibition receptor</td>\n",
       "      <td>KEGG: mmu04350</td>\n",
       "      <td>Secreted Signaling</td>\n",
       "      <td>Tgfb2  - (Tgfbr1+Tgfbr2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB3_TGFBR1_TGFBR2</th>\n",
       "      <td>TGFB3_TGFBR1_TGFBR2</td>\n",
       "      <td>TGFb</td>\n",
       "      <td>Tgfb3</td>\n",
       "      <td>TGFbR1_R2</td>\n",
       "      <td>TGFb agonist</td>\n",
       "      <td>TGFb antagonist</td>\n",
       "      <td></td>\n",
       "      <td>TGFb inhibition receptor</td>\n",
       "      <td>KEGG: mmu04350</td>\n",
       "      <td>Secreted Signaling</td>\n",
       "      <td>Tgfb3  - (Tgfbr1+Tgfbr2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB1_ACVR1B_TGFBR2</th>\n",
       "      <td>TGFB1_ACVR1B_TGFBR2</td>\n",
       "      <td>TGFb</td>\n",
       "      <td>Tgfb1</td>\n",
       "      <td>ACVR1B_TGFbR2</td>\n",
       "      <td>TGFb agonist</td>\n",
       "      <td>TGFb antagonist</td>\n",
       "      <td></td>\n",
       "      <td>TGFb inhibition receptor</td>\n",
       "      <td>PMID: 27449815</td>\n",
       "      <td>Secreted Signaling</td>\n",
       "      <td>Tgfb1  - (Acvr1b+Tgfbr2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB1_ACVR1C_TGFBR2</th>\n",
       "      <td>TGFB1_ACVR1C_TGFBR2</td>\n",
       "      <td>TGFb</td>\n",
       "      <td>Tgfb1</td>\n",
       "      <td>ACVR1C_TGFbR2</td>\n",
       "      <td>TGFb agonist</td>\n",
       "      <td>TGFb antagonist</td>\n",
       "      <td></td>\n",
       "      <td>TGFb inhibition receptor</td>\n",
       "      <td>PMID: 27449815</td>\n",
       "      <td>Secreted Signaling</td>\n",
       "      <td>Tgfb1  - (Acvr1c+Tgfbr2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIGIT_PVR</th>\n",
       "      <td>TIGIT_PVR</td>\n",
       "      <td>TIGIT</td>\n",
       "      <td>Tigit</td>\n",
       "      <td>Pvr</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PMID:24987108</td>\n",
       "      <td>Cell-Cell Contact</td>\n",
       "      <td>Tigit  - Pvr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITGA4_ITGB1_VCAM1</th>\n",
       "      <td>ITGA4_ITGB1_VCAM1</td>\n",
       "      <td>VCAM</td>\n",
       "      <td>ITGA4_ITGB1</td>\n",
       "      <td>Vcam1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KEGG: mmu04514</td>\n",
       "      <td>Cell-Cell Contact</td>\n",
       "      <td>Vcam1 - (Itga4+Itgb1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITGA9_ITGB1_VCAM1</th>\n",
       "      <td>ITGA9_ITGB1_VCAM1</td>\n",
       "      <td>VCAM</td>\n",
       "      <td>ITGA9_ITGB1</td>\n",
       "      <td>Vcam1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KEGG: mmu04514</td>\n",
       "      <td>Cell-Cell Contact</td>\n",
       "      <td>Vcam1 - (Itga9+Itgb1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITGA4_ITGB7_VCAM1</th>\n",
       "      <td>ITGA4_ITGB7_VCAM1</td>\n",
       "      <td>VCAM</td>\n",
       "      <td>ITGA4_ITGB7</td>\n",
       "      <td>Vcam1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KEGG: mmu04514</td>\n",
       "      <td>Cell-Cell Contact</td>\n",
       "      <td>Vcam1 - (Itga4+Itgb7)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VSIR_IGSF11</th>\n",
       "      <td>VSIR_IGSF11</td>\n",
       "      <td>VISTA</td>\n",
       "      <td>Vsir</td>\n",
       "      <td>Igsf11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KEGG: mmu04514</td>\n",
       "      <td>Cell-Cell Contact</td>\n",
       "      <td>Vsir  - Igsf11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2019 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        interaction_name pathway_name       ligand  \\\n",
       "TGFB1_TGFBR1_TGFBR2  TGFB1_TGFBR1_TGFBR2         TGFb        Tgfb1   \n",
       "TGFB2_TGFBR1_TGFBR2  TGFB2_TGFBR1_TGFBR2         TGFb        Tgfb2   \n",
       "TGFB3_TGFBR1_TGFBR2  TGFB3_TGFBR1_TGFBR2         TGFb        Tgfb3   \n",
       "TGFB1_ACVR1B_TGFBR2  TGFB1_ACVR1B_TGFBR2         TGFb        Tgfb1   \n",
       "TGFB1_ACVR1C_TGFBR2  TGFB1_ACVR1C_TGFBR2         TGFb        Tgfb1   \n",
       "...                                  ...          ...          ...   \n",
       "TIGIT_PVR                      TIGIT_PVR        TIGIT        Tigit   \n",
       "ITGA4_ITGB1_VCAM1      ITGA4_ITGB1_VCAM1         VCAM  ITGA4_ITGB1   \n",
       "ITGA9_ITGB1_VCAM1      ITGA9_ITGB1_VCAM1         VCAM  ITGA9_ITGB1   \n",
       "ITGA4_ITGB7_VCAM1      ITGA4_ITGB7_VCAM1         VCAM  ITGA4_ITGB7   \n",
       "VSIR_IGSF11                  VSIR_IGSF11        VISTA         Vsir   \n",
       "\n",
       "                          receptor       agonist       antagonist  \\\n",
       "TGFB1_TGFBR1_TGFBR2      TGFbR1_R2  TGFb agonist  TGFb antagonist   \n",
       "TGFB2_TGFBR1_TGFBR2      TGFbR1_R2  TGFb agonist  TGFb antagonist   \n",
       "TGFB3_TGFBR1_TGFBR2      TGFbR1_R2  TGFb agonist  TGFb antagonist   \n",
       "TGFB1_ACVR1B_TGFBR2  ACVR1B_TGFbR2  TGFb agonist  TGFb antagonist   \n",
       "TGFB1_ACVR1C_TGFBR2  ACVR1C_TGFbR2  TGFb agonist  TGFb antagonist   \n",
       "...                            ...           ...              ...   \n",
       "TIGIT_PVR                      Pvr                                  \n",
       "ITGA4_ITGB1_VCAM1            Vcam1                                  \n",
       "ITGA9_ITGB1_VCAM1            Vcam1                                  \n",
       "ITGA4_ITGB7_VCAM1            Vcam1                                  \n",
       "VSIR_IGSF11                 Igsf11                                  \n",
       "\n",
       "                    co_A_receptor             co_I_receptor        evidence  \\\n",
       "TGFB1_TGFBR1_TGFBR2                TGFb inhibition receptor  KEGG: mmu04350   \n",
       "TGFB2_TGFBR1_TGFBR2                TGFb inhibition receptor  KEGG: mmu04350   \n",
       "TGFB3_TGFBR1_TGFBR2                TGFb inhibition receptor  KEGG: mmu04350   \n",
       "TGFB1_ACVR1B_TGFBR2                TGFb inhibition receptor  PMID: 27449815   \n",
       "TGFB1_ACVR1C_TGFBR2                TGFb inhibition receptor  PMID: 27449815   \n",
       "...                           ...                       ...             ...   \n",
       "TIGIT_PVR                                                     PMID:24987108   \n",
       "ITGA4_ITGB1_VCAM1                                            KEGG: mmu04514   \n",
       "ITGA9_ITGB1_VCAM1                                            KEGG: mmu04514   \n",
       "ITGA4_ITGB7_VCAM1                                            KEGG: mmu04514   \n",
       "VSIR_IGSF11                                                  KEGG: mmu04514   \n",
       "\n",
       "                             annotation        interaction_name_2  \n",
       "TGFB1_TGFBR1_TGFBR2  Secreted Signaling   Tgfb1 - (Tgfbr1+Tgfbr2)  \n",
       "TGFB2_TGFBR1_TGFBR2  Secreted Signaling  Tgfb2  - (Tgfbr1+Tgfbr2)  \n",
       "TGFB3_TGFBR1_TGFBR2  Secreted Signaling  Tgfb3  - (Tgfbr1+Tgfbr2)  \n",
       "TGFB1_ACVR1B_TGFBR2  Secreted Signaling  Tgfb1  - (Acvr1b+Tgfbr2)  \n",
       "TGFB1_ACVR1C_TGFBR2  Secreted Signaling  Tgfb1  - (Acvr1c+Tgfbr2)  \n",
       "...                                 ...                       ...  \n",
       "TIGIT_PVR             Cell-Cell Contact              Tigit  - Pvr  \n",
       "ITGA4_ITGB1_VCAM1     Cell-Cell Contact     Vcam1 - (Itga4+Itgb1)  \n",
       "ITGA9_ITGB1_VCAM1     Cell-Cell Contact     Vcam1 - (Itga9+Itgb1)  \n",
       "ITGA4_ITGB7_VCAM1     Cell-Cell Contact     Vcam1 - (Itga4+Itgb7)  \n",
       "VSIR_IGSF11           Cell-Cell Contact            Vsir  - Igsf11  \n",
       "\n",
       "[2019 rows x 11 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r[\"load\"](\"CellChatDB.mouse.rda\")\n",
    "\n",
    "ccDB_mouse = r[\"CellChatDB.mouse\"]\n",
    "ccDB_mouse[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>EntrezGene.ID</th>\n",
       "      <th>Ensembl.Gene.ID</th>\n",
       "      <th>HomoloGene.ID</th>\n",
       "      <th>HGNC.ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGI:87853</th>\n",
       "      <td>a</td>\n",
       "      <td>nonagouti</td>\n",
       "      <td>50518</td>\n",
       "      <td>ENSMUSG00000027596</td>\n",
       "      <td>1264</td>\n",
       "      <td>HGNC:745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87854</th>\n",
       "      <td>Pzp</td>\n",
       "      <td>PZP, alpha-2-macroglobulin like</td>\n",
       "      <td>11287</td>\n",
       "      <td>ENSMUSG00000030359</td>\n",
       "      <td>104112</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87859</th>\n",
       "      <td>Abl1</td>\n",
       "      <td>c-abl oncogene 1, non-receptor tyrosine kinase</td>\n",
       "      <td>11350</td>\n",
       "      <td>ENSMUSG00000026842</td>\n",
       "      <td>3783</td>\n",
       "      <td>HGNC:76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87860</th>\n",
       "      <td>Abl2</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene 2...</td>\n",
       "      <td>11352</td>\n",
       "      <td>ENSMUSG00000026596</td>\n",
       "      <td>5278</td>\n",
       "      <td>HGNC:77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87862</th>\n",
       "      <td>Scgb1b27</td>\n",
       "      <td>secretoglobin, family 1B, member 27</td>\n",
       "      <td>11354</td>\n",
       "      <td>ENSMUSG00000066583</td>\n",
       "      <td>114479</td>\n",
       "      <td>null</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:1913728</th>\n",
       "      <td>Cd99</td>\n",
       "      <td>CD99 antigen</td>\n",
       "      <td>673094</td>\n",
       "      <td></td>\n",
       "      <td>48107</td>\n",
       "      <td>HGNC:7082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:96876</th>\n",
       "      <td>Klra</td>\n",
       "      <td>killer cell lectin-like receptor, subfamily A</td>\n",
       "      <td>17055</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HGNC:6372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:1891386</th>\n",
       "      <td>Ccl21c</td>\n",
       "      <td>chemokine (C-C motif) ligand 21C (leucine)</td>\n",
       "      <td>111654</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HGNC:10620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:107670</th>\n",
       "      <td>Ifna</td>\n",
       "      <td>interferon alpha</td>\n",
       "      <td>4663</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:109477</th>\n",
       "      <td>Raet1a</td>\n",
       "      <td>retinoic acid early transcript 1, alpha</td>\n",
       "      <td>19368</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45551 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Symbol                                               Name  \\\n",
       "MGI:87853           a                                          nonagouti   \n",
       "MGI:87854         Pzp                    PZP, alpha-2-macroglobulin like   \n",
       "MGI:87859        Abl1     c-abl oncogene 1, non-receptor tyrosine kinase   \n",
       "MGI:87860        Abl2  v-abl Abelson murine leukemia viral oncogene 2...   \n",
       "MGI:87862    Scgb1b27                secretoglobin, family 1B, member 27   \n",
       "...               ...                                                ...   \n",
       "MGI:1913728      Cd99                                       CD99 antigen   \n",
       "MGI:96876        Klra      killer cell lectin-like receptor, subfamily A   \n",
       "MGI:1891386    Ccl21c         chemokine (C-C motif) ligand 21C (leucine)   \n",
       "MGI:107670       Ifna                                   interferon alpha   \n",
       "MGI:109477     Raet1a            retinoic acid early transcript 1, alpha   \n",
       "\n",
       "             EntrezGene.ID     Ensembl.Gene.ID HomoloGene.ID     HGNC.ID  \n",
       "MGI:87853            50518  ENSMUSG00000027596          1264    HGNC:745  \n",
       "MGI:87854            11287  ENSMUSG00000030359        104112        null  \n",
       "MGI:87859            11350  ENSMUSG00000026842          3783     HGNC:76  \n",
       "MGI:87860            11352  ENSMUSG00000026596          5278     HGNC:77  \n",
       "MGI:87862            11354  ENSMUSG00000066583        114479        null  \n",
       "...                    ...                 ...           ...         ...  \n",
       "MGI:1913728         673094                             48107   HGNC:7082  \n",
       "MGI:96876            17055                                     HGNC:6372  \n",
       "MGI:1891386         111654                                    HGNC:10620  \n",
       "MGI:107670            4663                                                \n",
       "MGI:109477           19368                                                \n",
       "\n",
       "[45551 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccDB_mouse[3]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
